vmware-tanzu/tanzu-framework
# CLI: Add target separation

### What this PR does / why we need it

1. Currently, most stand-alone plugins are specific to the k8s context.
Hence, adding the context type to discovery source and grouping the
stand-alone plugins by context type. Only the stand-alone plugins
that are associated with a context type show up under the type.

2. Group contextual plugins by context type
Added kubernetes and TMC commands under root which further contain
plugins specific to those contexts. The plugins also show up under
root to maintain backwards compatibility.

### Which issue(s) this PR fixes
Fixes #2201

### Describe testing done for PR
Added unit tests for the config changes. Tested the `tanzu` targets  manually with the following steps:
1. Download an older version of the Tanzu CLI (`v0.21.0`)
2. Login with a management cluster `tanzu login`
3. Get management cluster `tanzu management-cluster get` & list clusters `tanzu cluster list` to ensure that the plugins are working
4. Upgrade to the new CLI with context command `make install-cli` and list the contexts `tanzu context list` to ensure the new CLI translates servers to contexts
5. Run `tanzu init` to update the new catalog fields
6. Get management cluster `tanzu management-cluster get` & list clusters `tanzu cluster list` to ensure that the existing invocations are still working
7. Get management cluster `tanzu kubernetes management-cluster get` & list clusters `tanzu kubernetes cluster list` to ensure that the new invocations are working

### Release note
```release-note
Currently, all CLI plugins are specific to the Kubernetes context. As more services like TMC add plugins, there is some room for confusion as there is name collision between the plugins. To make the user experience very explicit, we are adding the target to the command invocation.

With this change, the Tanzu CLI supports two targets (context types): `kubernetes`, `mission-control`. This is currently backwards compatible, i.e., the plugins are still available at the root level. In addition to that, we have all the contextual plugins grouped under their corresponding targets.

Note: This is currently behind a feature flag. To enable the flag please run tanzu config set features.global.context-target true and then run `tanzu init` once after the upgrade.

Please find some examples below:

1. TKG workload cluster plugin:
# Without target grouping (a TKG management cluster is set as the current active server)
`~ tanzu cluster`

# With target grouping
`~ tanzu kubernetes cluster`

2. TMC workload cluster plugin:
# Without target grouping (a TMC server is set as the current active server)
`~ tanzu cluster`

# With target grouping
`~ tanzu mission-control cluster`
```

### PR Checklist

### Additional information
